contrib/libpcap/utils/utils_libpcap.cc

```diff
@@ -0,0 +1,170 @@
+    api_.sapi_fclose(&pcap_pfile).IgnoreError();
+    sapi::v::RemotePtr remote_ptr_error(sapi_pcap_errmsg.GetRemote());
+    SAPI_ASSIGN_OR_RETURN(std::string errmsg,
+                          sandbox_->GetCString(remote_ptr_error));
+    return absl::UnavailableError(errmsg);
+  }
+
+  sapi_pcap_context_.SetRemote(pcap_context);
+  return absl::OkStatus();
+}
+
+LibPcap::~LibPcap() {
+  if (sapi_pcap_context_.GetRemote() != nullptr) {
+    api_.pcap_close(sapi_pcap_context_.PtrNone()).IgnoreError();
+  }
+}
+
+bool LibPcap::IsInit() { return CheckIsInit().ok(); }
+
+absl::Status LibPcap::CheckIsInit() {
+  if (fd_.GetValue() < 0) {
+    return absl::UnavailableError("PCAP file not opened");
+  }
+  if (!init_status_.ok()) {
+    return init_status_;
+  }
+
+  return absl::OkStatus();
+}
+
+absl::Status LibPcap::GetInitStatus() { return init_status_; }
+
+absl::StatusOr<LibPcapPacket> LibPcap::Next() {
+  SAPI_RETURN_IF_ERROR(CheckIsInit());
+
+  LibPcapPacket packet(&sapi_pcap_context_, &api_);
+  if (!packet.GetInitStatus().ok()) {
+    return packet.GetInitStatus();
+  }
+
+  return packet;
+}
+
+absl::Status LibPcap::SetFilter(const std::string& filter, int optimize,
+                                uint32_t netmask) {
+  sapi::v::Struct<struct bpf_program> sapi_bpf_program;
+  sapi::v::ConstCStr sapi_filter(filter.c_str());
+
+  int ret;
+  SAPI_ASSIGN_OR_RETURN(
+      ret, api_.pcap_compile(sapi_pcap_context_.PtrNone(),
+                             sapi_bpf_program.PtrAfter(),
+                             sapi_filter.PtrBefore(), optimize, netmask));
+  if (ret == -1) {
+    return absl::UnavailableError("Unable to compile filter");
+  }
+
+  SAPI_ASSIGN_OR_RETURN(ret, api_.pcap_setfilter(sapi_pcap_context_.PtrNone(),
+                                                 sapi_bpf_program.PtrNone()));
+  if (ret == -1) {
+    return absl::UnavailableError("Unable to set filter");
+  }
+
+  return absl::OkStatus();
+}
+
+absl::Status LibPcapPacket::FetchPackage(sapi::v::GenericPtr* context,
+                                         LibpcapApi* api) {
+  sapi::v::Struct<struct pcap_pkthdr> sapi_packet_header;
+  SAPI_ASSIGN_OR_RETURN(
+      uint8_t * package,
+      api->pcap_next(context->PtrNone(), sapi_packet_header.PtrAfter()));
+
+  if (package == nullptr) {
+    finished_ = true;
+    return absl::OkStatus();
+  }
+
+  memcpy(&packet_header_, &sapi_packet_header.data(), sizeof(packet_header_));
+
+  size_t bufsize = packet_header_.caplen;
+  if (bufsize > kMaxPacket) {
+    return absl::UnavailableError("Package to large");
+  }
+
+  buffer_.resize(bufsize);
+  sapi::v::Array<uint8_t> sapi_buffer(buffer_.data(), buffer_.size());
+  sapi_buffer.SetRemote(package);
+  SAPI_RETURN_IF_ERROR(api->GetSandbox()->TransferFromSandboxee(&sapi_buffer));
+
+  return absl::OkStatus();
+}
+
+absl::Status LibPcapPacket::GetInitStatus() const { return init_status_; }
+
+absl::StatusOr<const uint8_t*> LibPcapPacket::GetData() const {
```

Comment:
instead of separate data/size functions perhaps return `absl::StaturOr<absl::Span<const uint8_t>>`